p4-16/spec/P4-16-spec.mdk

```diff
@@ -6416,6 +6416,154 @@ ck16.update( { h.ipv4.ttl, h.ipv4.proto } );
+| value types | yes     | yes    | yes     | yes    |
+|-------------|---------|--------|---------|--------|
+
+The next table lists restrictions on where one may perform
+instantiations (see Section [#sec-instantiations]) of different types.
+The answer for `package` is always "no" because there is no
+"inside a package" where instantiations can be written in P4~16~.  One
+can definitely make constructor calls and use instances of stateful
+types as parameters when instantiating a package, and restrictions on
+those types are in the table above.
+
+For externs, one can only specify their interface in P4~16~, not their
+implementation.  Thus there is no place to instantiate objects within
+an extern.
+
+You may declare variables and constants of any of the value types
+within a parser or control (see Section [#sec-variables] for more
+details).  Declaring a variable or constant is not the same as
+instantiation, hence the answer "N/A" (for not applicable) in those
+table entries.  Variables may not be declared at the top level of your
+program, but constants may.
+
+|-------------|-------------------------------------------------|
+|             | can be instantiated in this place |||||
+|             |-----------|---------|---------|---------|--------|
+| This type   | top level | package | parser  | control | extern |
++-------------+:---------:+:-------:+:-------:+:-------:+:------:+
+| package     | yes       |  no     | no      | no      |  no    |
+| parser      |  no       |  no     | yes     | no      |  no    |
+| control     |  no       |  no     | no      | yes     |  no    |
+| extern      | yes       |  no     | yes     | yes     |  no    |
+| table       |  no       |  no     | no      | yes     |  no    |
+| value types | N/A       | N/A     |N/A      |N/A      | N/A    |
+|-------------|-----------|---------|---------|---------|--------|
+
+The next table lists restrictions on what types can be passed as
+run-time parameters to other callable things that have run-time
+parameters: parsers, controls, extern methods, and actions.
+
+|-------------|-------------------------------------------------|
+|             | can be a run-time parameter to this callable thing ||||
+|             |-----------|---------|---------|---------|
+| This type   | parser    | control | method  | action  |
++-------------+:---------:+:-------:+:-------:+:-------:+
+| package     |  no       |  no     | no      | no      |
+| parser      |  no       |  no     | no      | no      |
+| control     |  no       |  no     | no      | no      |
+| extern      | yes       | yes     | yes     | no      |
+| table       |  no       |  no     | no      | no      |
+| value types | yes       | yes     | yes     | yes     |
+|-------------|-----------|---------|---------|---------|
+
+The restrictions on return types of extern methods is the same as for
+parameters, with the addition that `void` can be a return type for
+methods that return no value.
```

Comment:
The table above and this sentence imply that it should be allowed to return an extern from an extern method call.  I'm not sure that is intended to be allowed?

If not, this sentence should probably be replaced with: "Extern method calls may return only value types, or nothing (specified by a return type of `void`)."

Reply:
Yes, let's start with the most restrictive language possible, and make it more lenient if there is a compelling use case. I think that functions should not be allowed to return extern types.

Reply:
Done.

Reply:
Do we want to distinguish constructor parameters from apply parameters here?

Reply:
I think we do, and I've tried to do so with these changes, but please suggest ways to make it clearer, if it is not.

There are currently 4 tables in these additions.  They are, in order:

(a) restrictions on constructor parameter types.  The table headings are "This type..." and "can be a constructor parameter for this type"

(b) restrictions on where in the source code instantiations are allowed.  The table headings are "This type" and "can be instantiated in this place"

(c) restrictions on run-time parameter types, which is what I think you mean by "apply parameters"?  The headings are "This type" and "can be a run-time parameter to this callable thing"

(d) restrictions on where calls to different kinds of callable things may occur.  The headings are "This type" and "can be called at run time from this place in a P4 program"

Reply:
Sorry, my fault. I was looking at GitHub diffs and missed it. Let me rebuild and check closely.